### ISSUE TITLE
Question: How to use even sized kernels with Monai Convolution block? 

### ISSUE THREAD

**Issue by gagandaroach** [Project-MONAI/MONAI#939]
Hi All, I am working on rewriting [MC_GAN](https://github.com/HYOJINPARK/MC_GAN/blob/master/Model1/model.py) with the MONAI framework and have a question with the Monai Convolution block. Why are even number kernels not supported in the Monai Convolution block?

I am running into this problem with the block when my kernel size is an even number. The `monai.networks.layers.convutils.same_padding()` func called by `Convolution().__init__`  is throwing a NotImplementedError:
```python
# from monai.networks.layers.convutils.same_padding()

    if np.any((kernel_size_np - 1) * dilation % 2 == 1):
        raise NotImplementedError(
            f"Same padding not available for kernel_size={kernel_size_np} and dilation={dilation_np}."
        )

    padding_np = (kernel_size_np - 1) / 2 * dilation_np
```
Is there an implementation planned for this in the future? Why is there no manual padding option in the Convolution block constructor? 

I did some value testing and padding_np is an integer when k_size is odd, and ends in 0.5 when k_size is even. 

For reference these are the `nn.Sequential()` blocks:
```python
# Original PyTorch source I am replacing:
            torch.nn.Conv2d(in_chan, out_chan, kernel_size=4, stride=2, padding=1, bias=False)
            torch.nn.BatchNorm2d(out_chan)
            torch.nn.LeakyReLU(0.2, inplace=True)
# What I want to use: Monai Convolution Block
            Convolution(2, in_chan, out_chan, kernel_size=4, strides=1, bias=False, act=self.Act, norm=self.Norm)
# What I am using: Monai LayerFactory
            Conv["conv", 2](in_chan, out_chan, 4, 2, 1, bias=False),
            Norm["batch", 2](out_chan),
            Act["leakyrelu"](0.2, inplace=True),
```

Thank you for any help.

**Comment by Nic-Ma** [comment]
Hi @ericspod ,

Could you please help share some suggestion here? I think you are expert on network layer topics.
Thanks in advance.

**Comment by wyli** [comment]
I remember there were some concerns about implementing a generic padding interface without knowing the input size, this will cause some performance/model serialisation issues

**Comment by ericspod** [comment]
The padding was the issue from how it was implemented originally. To pad for an even kernel size requires a slightly different pad scheme to ensure the output dimensions were the same as the input dimensions divided by the stride. With other values for stride, dilation, and whether the convolution was transpose or not, this wasn't easy to ensure. The intension of Convolution was to be used in places where a stride of 2 would produce a downsampling dividing an even input size in half which could then be upsampled again by a transpose convolution with a stride of 2, this is how the UNet class works. For other applications with kernel sizes that weren't odd we'd have to reformulate how Convolution works to get expected results, I was working on this by adding a padding layer that would add padding on one side of the input but it's trickier to get this to work as expected with dilation values other than 1 and with transpose convolutions. 

In your case Convolution as implemented will produce an output whose spatial dimensions are 1 less than the input, if you expect this or want to pad the input with 1 on one side of each spatial dimension that the class would work for you. Perhaps we should remove the exception check that is present in same_padding?

**Comment by gagandaroach** [comment]
Thanks @ericspod for the insightful response.

> In your case Convolution as implemented will produce an output whose spatial dimensions are 1 less than the input, if you expect this or want to pad the input with 1 on one side of each spatial dimension that the class would work for you. Perhaps we should remove the exception check that is present in same_padding?

I ran into the shrinking output issue yesterday when I added a padding param to `Convolution.__init__` and set it to 0. For my use case I think adding a `nn.functional.pad` step to the `nn.sequential block` before the MONAI `Convolution` is unnecessary when the `torch.nn.Conv2d` can handle the issue in one line with a manual padding param. If the exception check is removed the Convolution block may become more robust but I am not sure if that is its purpose when I can use a layer factory #836.

**Comment by ericspod** [comment]
The problem though is the padding for `Conv2d` or isn't sufficient to produce an output with the same shape given an even-sized kernel and a stride of 1:

```
t=torch.rand(1,1,64,64)

c=nn.Conv2d(1,1,4,1,padding=(0,0))
print(c(t).shape)  # 1x1x61x61

c=nn.Conv2d(1,1,4,1,padding=(1,1))
print(c(t).shape)  # 1x1x63x63

c=nn.Conv2d(1,1,4,1,padding=(2,2),)
print(c(t).shape)  # 1x1x65x65
```

Choosing to pad the input beforehand as well as pad in `Conv2d` can do this but when different stride or dilation values are used it becomes quite difficult to calculate what that pad should be. For example with a stride of 2 the expected output would be halved in every spatial dimension regardless of kernel size or dilation value, figuring out how to pad the input to do this isn't clear to me.

**Comment by gagandaroach** [comment]
Ok I am better understanding the problem posed. By default PyTorch's [`Conv2d`](https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html#torch.nn.Conv2d) drops the last column of the input when the k_size does not correlate with the output tensor. The MC GAN uses k_sizes of 4/8 because it is fixed for 128x128 size input. 

About the Convolution init and same_padding, I like the current exception policy. I do believe an override should exist for users looking for torch's functionality. For my project I discussed with the source researchers and concluded the kernel size can be changed to odd here.

**Comment by gagandaroach** [comment]
Is it possible to add functionality for no norm layer in the Convolution block?

**Comment by ericspod** [comment]
The argument `conv_only=True` will omit normalization, activation, and dropout from `Convolution`. To omit only normalization I've proposed the change [here](https://github.com/Project-MONAI/MONAI/pull/949).

**Comment by gagandaroach** [comment]
Thank you, this will be helpful.

The initial conv in MC GAN G/D does not use normalization but does use activation. Subsequent blocks use both. 

I think this issue can be closed now.
